archive/2025-11-08-code-execution-with-mcp-building-more-efficient-ai-agents-anthropic.md

```diff
@@ -0,0 +1,313 @@
+
+Cloudflare [published similar findings](https://blog.cloudflare.com/code-mode/), referring to code execution with MCP as “Code Mode." The core insight is the same: LLMs are adept at writing code and developers should take advantage of this strength to build agents that interact with MCP servers more efficiently.
+
+## **Benefits of code execution with MCP**
+
+Code execution with MCP enables agents to use context more efficiently by loading tools on demand, filtering data before it reaches the model, and executing complex logic in a single step. There are also security and state management benefits to using this approach.
+
+### Progressive disclosure
+
+Models are great at navigating filesystems. Presenting tools as code on a filesystem allows models to read tool definitions on-demand, rather than reading them all up-front.
+
+Alternatively, a `search_tools` tool can be added to the server to find relevant definitions. For example, when working with the hypothetical Salesforce server used above, the agent searches for "salesforce" and loads only those tools that it needs for the current task. Including a detail level parameter in the `search_tools` tool that allows the agent to select the level of detail required (such as name only, name and description, or the full definition with schemas) also helps the agent conserve context and find tools efficiently.
+
+### Context efficient tool results
+
+When working with large datasets, agents can filter and transform results in code before returning them. Consider fetching a 10,000-row spreadsheet:
+
+```
+// Without code execution - all rows flow through context
+TOOL CALL: gdrive.getSheet(sheetId: 'abc123')
+        → returns 10,000 rows in context to filter manually
+
+// With code execution - filter in the execution environment
+const allRows = await gdrive.getSheet({ sheetId: 'abc123' });
+const pendingOrders = allRows.filter(row =>
+  row["Status"] === 'pending'
+);
+console.log(`Found ${pendingOrders.length} pending orders`);
+console.log(pendingOrders.slice(0, 5)); // Only log first 5 for review
+```
+
+Copy
+
+The agent sees five rows instead of 10,000. Similar patterns work for aggregations, joins across multiple data sources, or extracting specific fields—all without bloating the context window.
+
+#### **More powerful and context-efficient control flow**
+
+Loops, conditionals, and error handling can be done with familiar code patterns rather than chaining individual tool calls. For example, if you need a deployment notification in Slack, the agent can write:
```

Comment:
This requires strong tests/compiler to catch and hint at programming mistakes early.